### PR TITLE
SONAR-7970 WS settings/set validates data

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 1]
```diff
@@ -97,6 +108,24 @@ public void handle(Request request, Response response) throws Exception {
...
+    PropertyDefinition.Result result = definition.validate(request.getValue());
+
+    checkRequest(result.isValid(),
+      i18n.message(Locale.ENGLISH, "property.error." + result.getErrorKey(), "Error when validating setting with key '%s' and value '%s'"),
+      request.getKey(), request.getValue());
+
+    checkRequest(component.isPresent() || definition.global(), "Setting '%s' cannot be global", request.getKey());
+    checkRequest(!component.isPresent()
+      || definition.qualifiers().contains(component.get().qualifier()),
+      "Setting '%s' cannot be set on a component with qualifier '%s'", request.getKey(), component.get().qualifier());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MAJOR Call "component.isPresent()" before accessing the value. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3655)

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 2]
```diff
@@ -97,6 +108,24 @@ public void handle(Request request, Response response) throws Exception {
     response.noContent();
   }
 
+  private void validate(SetRequest request, Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 3]
```diff
@@ -124,9 +154,12 @@ private static SetRequest toWsRequest(Request request) {
     return Optional.of(project);
   }
 
-  private static PropertyDto toProperty(SetRequest request, Optional<ComponentDto> component) {
+  private PropertyDto toProperty(SetRequest request, Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java [hunk 4]
```diff
@@ -97,6 +108,25 @@ public void handle(Request request, Response response) throws Exception {
...
+
+    checkRequest(result.isValid(),
+      i18n.message(Locale.ENGLISH, "property.error." + result.getErrorKey(), "Error when validating setting with key '%s' and value '%s'"),
+      request.getKey(), request.getValue());
+
+    checkRequest(component.isPresent() || definition.global(), "Setting '%s' cannot be global", request.getKey());
+    String qualifier = component.isPresent() ? component.get().qualifier() : "";
+    checkRequest(!component.isPresent()
+      || definition.qualifiers().contains(component.get().qualifier()),
+      "Setting '%s' cannot be set on a component with qualifier '%s'", request.getKey(), qualifier);
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java]: I'm not sure that displaying the qualifier from DB is a good idea... Will "TRK" talk to a user ?
Why not displaying the value from the bundle, for instance "qualifier.TRK" => "Project".
The message should then be changed to "Setting '%s' cannot be set on a '%s'".

**Comment by julienlancelot** [comment]
Please check my feedback and LGTM

**Pull request merged**